tracer/src/Datadog.Tracer.Native/cor_profiler.cpp

```diff
@@ -3337,9 +3337,16 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCachedFunctionSearchStarted(FunctionID
         return S_OK;
     }
 
-    // keep this lock until we are done using the module,
-    // to prevent it from unloading while in use
-    std::lock_guard<std::mutex> guard(module_ids_lock_);
+    try
+    {
+        // keep this lock until we are done using the module,
+        // to prevent it from unloading while in use
+        std::lock_guard<std::mutex> guard(module_ids_lock_);
+    }
+    catch (...)
```

Comment:
Not sure I understand the scenario here, could you please document it? 
When would an exception be thrown here, and why we are okay with ignoring it?

Reply:
I faced a process crash in one of our Exploration Tests, from the dump it was apparent that we failed to grab the mutex as it's already taken 'recursively' by the same thread (only in one of the OS's, not sure if it was Linux or Windows). The mutex `module_ids_lock_` is in heavy usage so I wanted to avoid changing its type to `recursive_mutex` so I surrounded everything with `try/catch`. We'll need to investigate it to the root to fix it appropriately.

Reply:
wow, we need to address this very soon.

Reply:
Totally, in the `catch` block I'm logging it as `Error` so if it will happen in the Ci, it will fail and we will be able to peek into the logs. This addition together with Omer's dump PR's will allow us to perform post-mortem analysis once one of the PRs will fail on that. We'll have to keep an eye on that.
I also opened a Jira ticket on myself to investigate it.